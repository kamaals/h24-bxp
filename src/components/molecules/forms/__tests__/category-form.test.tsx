import { describe, it, expect, vi, beforeEach } from "vitest";
import {
  render,
  screen,
  cleanup,
  fireEvent,
  waitFor,
} from "@testing-library/react";
import CategoryForm from "../category-form";
import ReduxStoreProvider from "@/components/providers/redux-store-provider";

const mockMutate = vi.fn();

vi.mock("@/lib/store/api/categoryService", () => ({
  useCreateCategoryMutation: vi.fn(() => [mockMutate]),
}));

describe(CategoryForm.name, async () => {
  beforeEach(() => {
    cleanup();
  });

  it("should render the form", () => {
    render(
      <ReduxStoreProvider user={null}>
        <CategoryForm />
      </ReduxStoreProvider>,
    );

    const form = screen.getByTestId("category-form");
    expect(form).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    const resp = render(
      <ReduxStoreProvider user={null}>
        <CategoryForm />
      </ReduxStoreProvider>,
    );

    expect(resp).matchSnapshot();
  });

  it("should render with parentId and data", () => {
    render(
      <ReduxStoreProvider user={null}>
        <CategoryForm
          edit={true}
          data={{ name: "T-Shits", id: "ffd" }}
          parentId={"0645"}
        />
      </ReduxStoreProvider>,
    );

    const form = screen.getByTestId("category-form");
    expect(form).toBeInTheDocument();
  });

  it("should call the api", async () => {
    const { container } = render(
      <ReduxStoreProvider user={null}>
        <CategoryForm
          edit={true}
          data={{ name: "T-Shits", id: "ffd" }}
          parentId={"0645"}
        />
      </ReduxStoreProvider>,
    );

    const button = screen.getByTestId("loading-button");

    await waitFor(() => {
      const nameInput = container.querySelector(
        'input[value="T-Shits"]',
      ) as HTMLInputElement;
      expect(nameInput).toBeInTheDocument();
      fireEvent.click(button);
      const svg = container.querySelector(".animate-spin");
      expect(svg).toBeInTheDocument();
    });
  });
});
