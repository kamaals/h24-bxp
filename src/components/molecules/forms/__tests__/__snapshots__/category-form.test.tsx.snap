// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoryForm > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="max-w-lg"
        data-testid="category-form"
      >
        <form>
          <div>
            <div
              class="grid grid-cols-4 gap-2"
              data-slot="form-item"
            >
              <div>
                <label
                  class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive"
                  data-error="false"
                  data-slot="form-label"
                  for="«r1»-form-item"
                >
                  Name
                </label>
                <p
                  class="text-muted-foreground text-xs"
                  data-slot="form-description"
                  id="«r1»-form-item-description"
                  style="margin-top: -.125rem;"
                />
              </div>
              <div
                aria-describedby="«r1»-form-item-description"
                aria-invalid="false"
                class="col-span-3 space-y-1"
                data-slot="form-control"
                id="«r1»-form-item"
              >
                <input
                  class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
                  data-slot="input"
                  data-testid="name"
                  name="name"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="py-4 flex justify-end animate"
          >
            <button
              class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
              data-slot="button"
              data-testid="loading-button"
              type="submit"
            >
              <span
                class="font-bold"
              >
                Submit
              </span>
            </button>
          </div>
        </form>
        <div>
          <div
            class="p-4"
          >
            <h5>
              Errors
            </h5>
            <pre>
              {}
            </pre>
          </div>
          <pre>
            {}
          </pre>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="max-w-lg"
      data-testid="category-form"
    >
      <form>
        <div>
          <div
            class="grid grid-cols-4 gap-2"
            data-slot="form-item"
          >
            <div>
              <label
                class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive"
                data-error="false"
                data-slot="form-label"
                for="«r1»-form-item"
              >
                Name
              </label>
              <p
                class="text-muted-foreground text-xs"
                data-slot="form-description"
                id="«r1»-form-item-description"
                style="margin-top: -.125rem;"
              />
            </div>
            <div
              aria-describedby="«r1»-form-item-description"
              aria-invalid="false"
              class="col-span-3 space-y-1"
              data-slot="form-control"
              id="«r1»-form-item"
            >
              <input
                class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
                data-slot="input"
                data-testid="name"
                name="name"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="py-4 flex justify-end animate"
        >
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
            data-slot="button"
            data-testid="loading-button"
            type="submit"
          >
            <span
              class="font-bold"
            >
              Submit
            </span>
          </button>
        </div>
      </form>
      <div>
        <div
          class="p-4"
        >
          <h5>
            Errors
          </h5>
          <pre>
            {}
          </pre>
        </div>
        <pre>
          {}
        </pre>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
